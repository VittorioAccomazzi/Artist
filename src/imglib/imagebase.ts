import { textChangeRangeIsUnchanged } from "typescript";


interface TypedArray {
    readonly length : number;
    [n: number]: number;
    buffer : ArrayBuffer;
}

interface TypedArrayConstructor<T extends TypedArray> {
    new (buffer: ArrayBuffer, byteOffset?: number, length?: number): T;
}

export type ImageUint8  = ImageBase<Uint8Array,Uint8ArrayConstructor>
export type ImageUint16 = ImageBase<Uint16Array,Uint16ArrayConstructor>
export type ImageFloat32= ImageBase<Float32Array,Float32ArrayConstructor>

export class ImageFactory {

    static Uint8( width : number, height : number ) : ImageUint8 {
        return new ImageBase<Uint8Array,Uint8ArrayConstructor>(Uint8Array.BYTES_PER_ELEMENT, Uint8Array, width, height)
    }
    static Uint16( width : number, height : number ) : ImageUint16{
        return (new ImageBase<Uint16Array,Uint16ArrayConstructor>(Uint16Array.BYTES_PER_ELEMENT, Uint16Array, width, height)) 
    }
    static Float32( width : number, height : number ): ImageFloat32{
        return new ImageBase<Float32Array,Float32ArrayConstructor>(Float32Array.BYTES_PER_ELEMENT, Float32Array, width, height)
    }
}


class ImageBase<T extends TypedArray, C extends TypedArrayConstructor<T>> {

    private buffer : T
    private pixels : T [] 

    constructor( elementSize : number, constructor : C, width : number, height : number ) {
        let nPixels= width * height;
        let bytes  = nPixels * elementSize
        let buffer= new ArrayBuffer(bytes);
        this.pixels= Array.from(Array(height), (e,i)=> new constructor (buffer, width*elementSize*i, width) ) 
        this.buffer= new constructor(buffer, 0, nPixels)
    }


    /**
     * get type of pixel stored.
     */
    get imageType(): string {
        return this.pixels[0].constructor.name.replace('Array','')
    }

    /**
     * returns the current list of pixels.
     */
    get imagePixels(): T {
        return this.buffer 
    }

    /**
     *  get the image buffer
     */
    get imageBuffer() : ArrayBuffer {
        return this.buffer.buffer
    }

    /**
     * get image height
     */
    get height() : number {
        return this.pixels.length
    }

    /**
     * return image height
     */
    get width() : number {
        return this.pixels[0].length
    }

    /**
     * get a pixel value.
     * @param x row of the pixel queried
     * @param y column of the pixel queried
     */
    get( x: number, y: number ) : number {
        if( !this.checkBoundary(x,y)) throw new Error (`Invalid image coordinates queried: image size ${this.width}x${this.height} pixel ${x},${y}`)
        return this.pixels[y][x]
    }

    /**
     * get a row of pixels
     * @param y row of pixels to select.
     */
    getRow(y:number) : T {
        if( y<0 || y>this.height )throw new Error (`Invalid y coordinate queried : image height ${this.height} y ${y}`)
        return this.pixels[y]
    }

    /**
     * 
     * @param x row pixel to set
     * @param y column pixel to se
     * @param value  value to set
     */
    set(x:number, y:number, value : number) : void {
        if( !this.checkBoundary(x,y)) throw new Error (`Invalid image coordinates set: image size ${this.width}x${this.height} pixel ${x},${y}`)
        this.pixels[y][x]= value
    }

    /**
     * check that the value passed are valid image coordinates.
     * @param x row
     * @param y column
     */
    private checkBoundary(x: number, y : number ) : boolean {
        return x>=0 && x < this.width && y>=0 && y<this.height
    }

}