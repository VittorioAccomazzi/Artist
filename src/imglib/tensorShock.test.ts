import { Canvas } from "canvas"
import CanvasUtils from "./canvasUtils"
import GaussianFilter from "./gaussianFilter"
import { Image2D, ImageFactory, ImageFloat32 } from "./imagebase"
import TensorGenerator from "./tensorGenerator"
import TensorShock from './tensorShock'
import { dumpCanvas, getCanvases, overlayTensor, toCanvas, toNodeCanvas, toSeqCanvas } from "./testutils"
import {hash} from './testutils'

test('shall smooth enhance edges',async()=>{

    for( const [image, name] of sampleImages()){
        let tf = TensorGenerator.Run(image, image, image, 1)
        let tI = overlayTensor(toCanvas(image), tf, 2)
        // await dumpCanvas(tI, `input tensor ${name}`)
        let hsh = await hash(tI)
        expect(hsh).toMatchSnapshot()
    
        let filtered = TensorShock.Run(image, tf, 1.5, 0.01, 2 )
        tI = overlayTensor(toCanvas(filtered), tf, 2)
        //await dumpCanvas(tI, `filtered with tensor Smoothing ${name}`)
        hsh = await hash(tI)
        expect(hsh).toMatchSnapshot()
    }

},30*1000)

function sampleImages() : [Image2D, string ][]{
    let width = 120
    let height= 180
    let bck = 50
    let frg = 100
    let image1= ImageFactory.Uint8(width,height)
    let pixels= image1.imagePixels
    pixels.forEach((v,i)=>{
        let y = (i/width | 0) 
        let x = (i%width)
        let val = bck
        if( x< y ) val = frg
        pixels[i]=val
    })
    GaussianFilter.Run(image1,2)
    let image2= ImageFactory.Uint8(width,height)
    pixels= image2.imagePixels
    pixels.forEach((v,i)=>{
        let y = (i/width | 0) 
        let x = (i%width)
        let val = bck
        if( Math.abs(x-y)<3 ) val = frg + (x/3|0)
        pixels[i]=val
    })

    return [[image1,'test1'],[image2,'test2']]
}

test('shall enhance edges',async()=>{
    for await ( const [canvas,ctx,name] of getCanvases() ) {
        let [lImg, aImg, bImg] = CanvasUtils.toLab(toSeqCanvas(canvas))
        
        GaussianFilter.Run(lImg,1)
        let tf = TensorGenerator.Run(lImg, aImg, bImg, 1) 
        let flt= TensorShock.Run(lImg, tf, 2, 0.01, 2) as ImageFloat32
        let cns= CanvasUtils.fromLab(flt, aImg, bImg)
        // await dumpCanvas( toNodeCanvas(cns), `tensor Shock ${name}`)
        let hsh = await hash(flt)
        expect(hsh).toMatchSnapshot()
    }
},60*2*1000)