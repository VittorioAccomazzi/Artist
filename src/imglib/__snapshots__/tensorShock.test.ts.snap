// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shall enhance edges 1`] = `"00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000c00000000000000000000000000000040591c000000000000000000000000003fff828000000000000000000000010fffffff800000000000000000000004bffffffdfe8000000000000000000003fffffffdfda0000000000000000000ffffffffffff80000000000000000027ffffffffffffff08000000000000003ffffffffffffffff400000000000005fffffffffffffffffc000000000000007fffffffffffffffffd004000000002ffffffffffffffffffffe88000000007ffffffffffffffffffffe3000000002ffffffffffffffffffffffe40000006bfffffffffffffffffffffff04400001ffffffffffffffffffffffffff2400bfffffffffffffffffffffffffffd200fffffffffffffffffffffffffffffc10ffffffffffffffffffffffffffffff80ffffffffffffffffffffffffffffffc0ffffffffffffffffffffffffffffffe0fffffffffffffffffffffffffffffff8fffffffffffffffffffffffffffffffcdffffffffffffffffffffffffffffffeffffffffffffffffffffffffffffffff00000000000000000000000000000000000000000000000800000000000000000000000000002874c000000000000000000000000003ffffda0000000000000000000000003ffffffe5f00000000000000000000007fffffffc41800000000000000000007fffffffffe7c0000000000000000000fffffffffff700000000000000000007fffffffffff00000000000000000001ffffffffffff80000000000000000001ffffffffffffe0000000000000000013fffffffffffff000000000000000000ffffffffffffffe00000000000000003ffffffffffffffa0000000000000001ffffffffffffffff8000000000000007fffffffffffffffff800000000000007fffffffffffffffffdc000000000001fffffffffffffffffffd000000000007ffffffffffffffffffff40000000003fffffffffffffffffffffe0000000003fffffffffffffffffffffff00000000ffffffffffffffffffffffffc0000001ffffffffffffffffffffffffe000000bfffffffffffffffffffffffff000001fffffffffffffffffffffffffff0001ffffffffffffffffffffffffffff8007ffffffffffffffffffff7ffff012000fffffffffffffffffffe6201b0000000fffffffffffffffffffa000000000000ffffffffffffffffff80000000000000fffffffffffdfc000000000000000000fffffffffff8d0000000000000000000fffffffffffffbffdfa0000000000000fffffffffffd80fbf47000009f800000fffffffffffc005e604c000000000000ffffffffff001fe00000000000000000ffffc00000000000000000000000000000000000000000000000000000000000000000000000005ffffffffe000c000000000000000003ffefffffffffffffff0000000112effffffffffffffffffffb000000003ffffffffffffdfffffffffb0000001ffffffffffffffffffbffff5f00000007bffffffffffffffffc43ffff00000067fffffffffffffffffffffff3301e1ffffffffffffffffffffe783fde7001fffffffffffffffffffff881860003fffffffffffffffffffc7800000000ffffffffffffffffffbffffd00000000bffffffffffffffffffffe0000000000ffffd9ffffffffffffffc20000000000fffffffffffffffffffffff780000000fff7fffffffffffffffec00000000000fffffff3e3ff9bfffffffc4000000000fdefffffffc0fffc01ff603000000000bfcbffffe03ff80ffff8000000000000dffffe001fe0dfff8020708200000000e1e0137ff87ffc3000300000000000001017ffc0ffff060e7e9f20000000000000006207ff20407ff000000000000000cf8407780000344000000000000000000007fe8007ffffc6007ff0000000000000780006fff000000e000000000000000384198ffa000000e070000000000000ffff7fffffffffffffffffffffffffefffffffffffffffffec3ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff1ffffffffffffe4ffffffffffffffffc3fffffffffffffffff3ffffe07fffffffffffffffffffffffffffffffffffffffffffffffffe0000c7afffffdfffffff9ffffffffe07ffffffffffe781fffffefffdff801fffffffe07fffffffffbfffffffffffffa7000700087ffffbffffe17ffffffcffffffffb9e7f0fe1e7e08cf9ffffffffcfa00003cfffffee780fffffffff3ffc00000500fffe001ffffffa03180e1fce1ffffe0c00006de1fe1ffffffffffb0047ff81c3c000083ff00001f0f3d0040e1000007dfd0003800fe7fffffffe2000300338000051000000180000fff83efffffe03fa0100020ef8000000ffffc00000100001ec00000187b8030f000004000700780000001cc07ffffff0000000000003c00000004200017ff8064000000061fc0000300000ff8e106000e30080efbfc0000000000007800000001ef00019e00000000000000000000000ff80000000000000000000000000000080000020000000000000000000000000000f8300000000000000000200000000fffc086300000000000000000000000f000080180000000000000000000000000000000000000000000000000000000000000000000000000000000000000000"`;

exports[`shall enhance edges 2`] = `"fffe0000b5ffffffb5007fffffc00003fffe0000adfffedf80003fffffc00000fffe00001fffeedde0003fffffe00000fffe000019ffffdfe0401fffffe00030fffe00003fffff5fecc01ffffff000e0fffe00007ffdffcee8001ffffff000c0fffe000015bbc003c0001ffffff80180fffe000003f9c008c0001ffffffc0300fffe000001b4c00080001ffffffc0600fffe0000004c8000c0001ffffffe0400fffe00000000800184000ffffffe0000fffe000004388bf800000fffffff0000fffe0000031cffff02000fffffff0000fffe00001989ffffc0000fffffff8000f9fe00000079ffffe0000fffffffc000f9fe000001f13ffff8000fffffff8000f1fe000001013ffffc000fffffff8000f1fe000003016ffffe000fffffff0000e1fe00000600dfffff800ffffffe0000e1fe000006007fffffc00fc7fffc0001c1fe0000040057ffffe00fe3fff8000381fe000008007ffffff00fe3fff0000381fe000000007ffffff00fe1fff0000701fe00000000fffffff80fe0ffe0000f01fe00000000fffffffc0fe0ffc0001f01fe00004009ffffffff0fe07fc0003f01fe0000c0077cffffff0fe07f80007f01fe0000c00671ffffff8fe01f0000ff01fe0000800ec3ffffffcfe03f0001ff01fe0001801c8fffffffcfe07f0001ff01fe000180031fffffffcfe0ff0003ff01fe000180443fffffffe7e3ff8007ff01fe000180007ffffffff7c7ff8007ff01fe00030000ffffffffff9fff000fff01fe00030001ffffffffffffff000fff01fe00030003dfffffffffffff001fff01fe00030007bfffffffffffff003fff00fe0003800f7fffffffffffff003fff01fe0003c01ceffffffffffffe007fff01fe0003e029fffffffffffffe007fff01fe0003c053affffffffffffe00ffff01fe0003d08797fffffffffffc01ffff01fe0007f00a2671ffffffffcc01ffff01fe0007f808480c7e7fffff0c01ffff01fe0003f810100e007ffffe1803ffff01fe0003c0202000007ffff03003ffff01fe0003c840400001ffffc0e007ffff01fe0003d800c00003ffff03c007ffff01fe0001e800400005fffc3f800fffff01fe0001f00000000bfff03f000fffff01fe0001f000000007fff03e0003ffff01fe0000f00000000ffff01c0007ffff00fe0000f00000001ffff0100007ffff01fe0000700000003ffff8300007ffff01fe0000700000007ffff830000fffff01fe000060000001bffff830000fffff00fe000860000003bffffc10005fffff00fe00006000000ffffffc10001fffff01fe000020000017bffffc1000bfffff00fe00000000000ffffffc1000bfffff00fe0000f000001feffffc10003fffff00fe00018800003fc1fffc10007fffe000fe0001080000ffb07ff018007ffff800fe0003800000fe483fc018007ffffc01ff000fe00001fc00bfc00803ffffff00ff001f800003f800bf000803ffffff01ff001e800003f00c3f080807fffff000ff000e004003e00c1e080807ffffe300ff001c004007c67e3e780807ffffef00ff02700000079f873fc0080fffffff00ff03c000001f1fff7fc0080ffffeff00ff010000000e3ffffff0080ffffdff00ff03e000403e3ffffff8080ffffdff00ff0f0000101c1ffffffc0c1ffff3ff00ff00000010785fffcffc0c1ffff7ff00ff00000008383fffcffc0c1fffefff00ff04000020b03fffcffc0c3fffefff00ff18010030603fffc7f8043fffdfff00ff00010000e03ffe67f8043fffdfff00ff00000002c03ffe77f8063fffbfff00ff00000003803ffe83f0067fffbfff01ff00008007001fff03f0067fffffff01ff0000440b001fffefe0067fffffff01ff04004072000fffffe0067fffffff01ff08002074003fffffe002ffff7fff01ff10000064003fffffc002ffff7fff01ff000013c8003ffffe4002ffff7fff01ff080005d00037c0008002ffff7fff01fe480407b00007e0ff8003fffe7fff01ff90040d600003ffff0003fffe7fff01ff10040fc00003f82f0003fffe7fff01ff200823800001fe0e0003fffe7fff01ff200903800000fffc0003fffe7fff01ff000913000001fffc0003fffe7fff01ff000888000001fffc0003fffe7fff01ff0000280000003ff80003fffefffe01ff300c0f0000001ff80003bffefff801ffd0060d800007fffc000303feffe001ff800664c00007ffff8003c07effc000ff200241e0000fffffe003f800ff80c0ff602343a00007fffff803fe04ff00e0ffc10002100007fffffc03ffc4fe00f8ff800107000007fffffe03ffedfc0078ffc00003280005ffffff07fffffc003cff800003980047ffffff07fffffc003cff802001c80047ffffff87fffff8003cff800001880007ffffffc7fffff8003cff800001c80007ffffffc7fffff0003cff8040099800c7ffffffc7fffff0003cff000009f001c7ffffffe7ffffe0001cff8000073001cfffffffe7ffffe0001cff8000038181cfffffffe7fffff0001cff8000004003cfffffffe7ffffe0001cff80000720038ffffffff7ffffc0001cff80000130039ffffffffdffffc0001cff80000010079ffffffff81fff80001cff00000800071ffffffff803ff80001e7f00000c000e3ffffffff800ff00001e7f000006000e7ffffffff8007f00001e7f000003401c7ffffffffc007e00001e7e0000008010fffffffffc007c00001e7e0000000001fffffffffc007800000e7f0000010003fffffffffe007000000e7f000001001ffffffffffe000000000e7f000000007ffffffffffe020000c00f7e000000007ffffffffffe078801800f7f000000007fffffffffff07b803000f7e00000000ffffffffffff07f80300"`;

exports[`shall smooth enhance edges 1`] = `"fff80000000000000000000000000000fffe0000000000000000000000000000ffff8000000000000000000000000000ffff8000000000000000000000000000ffffe000000000000000000000000000fffff800000000000000000000000000fffffa00000000000000000000000000fffffe00000000000000000000000000ffffff80000000000000000000000000ffffffb0000000000000000000000000fffffff0000000000000000000000000fffffffc000000000000000000000000ffffffff000000000000000000000000ffffffff000000000000000000000000ffffffffc00000000000000000000000fffffffff00000000000000000000000fffffffff40000000000000000000000fffffffffc0000000000000000000000ffffffffff0000000000000000000000ffffffffff6000000000000000000000ffffffffffe000000000000000000000fffffffffff800000000000000000000fffffffffffe00000000000000000000fffffffffffe00000000000000000000ffffffffffff80000000000000000000ffffffffffffe0000000000000000000ffffffffffffe8000000000000000000fffffffffffff8000000000000000000fffffffffffffe000000000000000000fffffffffffffe800000000000000000ffffffffffffff800000000000000000fffffffffffffff00000000000000000ffffffffffe000000000000000000000fffffffffde800000000000000000000ffffffffffaa00000000000000000000ffffffffffea00000000000000000000fffffffffffa80000000000000000000fffffffffffaa0000000000000000000fffffffffffea0000000000000000000fffffffffffff8000000000000000000ffffffffffffaa000000000000000000ffffffffffffe1000000000000000000fffffffffffffd400000000000000000fffffffffffffaf00000000000000000fffffffffffffe540000000000000000ffffffffffffffd40000000000000000fffffffffffffffd0000000000000000fffffffffffffff54000000000000000fffffffffffffffd4000000000000000fffffffffffffffff800000000000000ffffffffffffffff5400000000000000ffffffffffffffffd000000000000000fffffffffffffffffa80000000000000fffffffffffffffff5c0000000000000fffffffffffffffffca0000000000000ffffffffffffffffffa8000000000000fffffffffffffffffffe000000000000ffffffffffffffffffea800000000000fffffffffffffffffffa800000000000fffffffffffffffffffff00000000000fffffffffffffffffffea80000000000ffffffffffffffffffffa00000000000fffffffffffffffffffff50000000000ffffffffffffffffffffea800000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000"`;

exports[`shall smooth enhance edges 2`] = `"fff80000000000000000000000000000fffe0000000000000000000000000000ffff8000000000000000000000000000ffff8000000000000000000000000000ffffe000000000000000000000000000fffff800000000000000000000000000fffffa00000000000000000000000000fffffe00000000000000000000000000ffffff80000000000000000000000000ffffffb0000000000000000000000000fffffff0000000000000000000000000fffffffc000000000000000000000000ffffffff000000000000000000000000ffffffff000000000000000000000000ffffffffc00000000000000000000000fffffffff00000000000000000000000fffffffff40000000000000000000000fffffffffc0000000000000000000000ffffffffff0000000000000000000000ffffffffff6000000000000000000000ffffffffffe000000000000000000000fffffffffff800000000000000000000fffffffffffe00000000000000000000fffffffffffe00000000000000000000ffffffffffff80000000000000000000ffffffffffffe0000000000000000000ffffffffffffe8000000000000000000fffffffffffff8000000000000000000fffffffffffffe000000000000000000fffffffffffffe800000000000000000ffffffffffffff800000000000000000fffffffffffffff00000000000000000fffffffffea800000000000000000000fffffffffd0a00000000000000000000ffffffffff2a80000000000000000000ffffffffffea80000000000000000000fffffffffffaa0000000000000000000fffffffffffaa0000000000000000000fffffffffffe84000000000000000000ffffffffffffaa000000000000000000ffffffffffffa8000000000000000000ffffffffffffe0500000000000000000fffffffffffffd500000000000000000fffffffffffffa940000000000000000fffffffffffffe150000000000000000ffffffffffffffd50000000000000000ffffffffffffffb54000000000000000fffffffffffffff54000000000000000fffffffffffffffd4000000000000000ffffffffffffffff5e00000000000000ffffffffffffffff5400000000000000ffffffffffffffffc000000000000000fffffffffffffffffaa0000000000000fffffffffffffffff548000000000000fffffffffffffffffc2a000000000000ffffffffffffffffffaa000000000000ffffffffffffffffff6a800000000000ffffffffffffffffffcaa00000000000fffffffffffffffffffa800000000000fffffffffffffffffffeb40000000000fffffffffffffffffffeaa0000000000ffffffffffffffffffff800000000000fffffffffffffffffffff54000000000ffffffffffffffffffffeab00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000"`;

exports[`shall smooth enhance edges 3`] = `"ffe00000000000000000000000000000fff80000000000000000000000000000fffe0000000000000000000000000000fffe0000000000000000000000000000ffff8000000000000000000000000000ffffe000000000000000000000000000ffffe8000000000000000000000000007ffff8000000000000000000000000002ffffe000000000000000000000000000ffffe8000000000000000000000000003ffff8000000000000000000000000000fffff000000000000000000000000000fffffc000000000000000000000000003ffffc000000000000000000000000000fffff000000000000000000000000000fffffc000000000000000000000000003ffffd000000000000000000000000000fffff0000000000000000000000000005ffffc000000000000000000000000003ffffd0000000000000000000000000007ffff0000000000000000000000000001ffffe000000000000000000000000001fffff8000000000000000000000000007ffff0000000000000000000000000001ffffe000000000000000000000000001fffff8000000000000000000000000007ffffa000000000000000000000000001ffffe000000000000000000000000000bffff8000000000000000000000000006ffffa000000000000000000000000000ffffe0000000000000000000000000003ffff8000000000000000000000000002ffffb000000000000000000000000000ffffec000000000000000000000000003ffffc000000000000000000000000002fffff000000000000000000000000000fffffc000000000000000000000000003ffffc0000000000000000000000000007ffff000000000000000000000000000fffffc000000000000000000000000003ffffc0000000000000000000000000007ffff0000000000000000000000000005ffffe000000000000000000000000001ffffd0000000000000000000000000007ffff8000000000000000000000000005ffffe000000000000000000000000001fffff8000000000000000000000000007ffff8000000000000000000000000000ffffe000000000000000000000000001fffff8000000000000000000000000007ffff8000000000000000000000000000ffffe000000000000000000000000000bffff8000000000000000000000000003ffffa000000000000000000000000000fffff000000000000000000000000000bffffc000000000000000000000000003fffff000000000000000000000000000fffff0000000000000000000000000003ffffc000000000000000000000000003fffff000000000000000000000000000fffff0000000000000000000000000003ffffc0000000000000000000000000037ffff0000000000000000000000000007ffff4000000000000000000000000001ffffe0000000000000000000000000007ffff8000000000000000000000000007ffffe000000000000000000000000001ffffe0000000000000000000000000007ffff8000000000000000000000000007ffffe000000000000000000000000001ffffe8000000000000000000000000007ffff8000000000000000000000000002ffffe000000000000000000000000000ffffe8000000000000000000000000003ffff8000000000000000000000000000fffff000000000000000000000000000fffffc000000000000000000000000003ffffc000000000000000000000000000fffff000000000000000000000000000fffff0000000000000000000000000003ffff0000000000000000000000000000ffff00000000000000000000000000005fff00000000000000000000000000003fff000000000000000000000000000007ff000000000000000000000000000001ff000000000000000000000000000001ff0000000000000000000000000000007f0000000000000000000000000000001f0000000000000000000000000000001e0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000"`;

exports[`shall smooth enhance edges 4`] = `"ffe00000000000000000000000000000fff80000000000000000000000000000fffe0000000000000000000000000000fffe0000000000000000000000000000ffff8000000000000000000000000000ffffe000000000000000000000000000ffffe8000000000000000000000000007ffff8000000000000000000000000002ffffe000000000000000000000000000ffffe8000000000000000000000000003ffff8000000000000000000000000000fffff000000000000000000000000000fffffc000000000000000000000000003ffffc000000000000000000000000000fffff000000000000000000000000000fffffc000000000000000000000000003ffffd000000000000000000000000000fffff0000000000000000000000000005ffffc000000000000000000000000003ffffd0000000000000000000000000007ffff0000000000000000000000000001ffffe000000000000000000000000001fffff8000000000000000000000000007ffff0000000000000000000000000001ffffe000000000000000000000000001fffff8000000000000000000000000007ffffa000000000000000000000000001ffffe000000000000000000000000000bffff8000000000000000000000000006ffffa000000000000000000000000000ffffe0000000000000000000000000003ffff8000000000000000000000000002ffffb000000000000000000000000000ffffec000000000000000000000000003ffffc000000000000000000000000002fffff000000000000000000000000000fffffc000000000000000000000000003ffffc0000000000000000000000000007ffff000000000000000000000000000fffffc000000000000000000000000003ffffc0000000000000000000000000007ffff0000000000000000000000000005ffffe000000000000000000000000001ffffd0000000000000000000000000007ffff8000000000000000000000000005ffffe000000000000000000000000001fffff8000000000000000000000000007ffff8000000000000000000000000000ffffe000000000000000000000000001fffff8000000000000000000000000007ffff8000000000000000000000000000ffffe000000000000000000000000000bffff8000000000000000000000000003ffffa000000000000000000000000000fffff000000000000000000000000000bffffc000000000000000000000000003fffff000000000000000000000000000fffff0000000000000000000000000003ffffc000000000000000000000000003fffff000000000000000000000000000fffff0000000000000000000000000003ffffc0000000000000000000000000037ffff0000000000000000000000000007ffff4000000000000000000000000001ffffe0000000000000000000000000007ffff8000000000000000000000000007ffffe000000000000000000000000001ffffe0000000000000000000000000007ffff8000000000000000000000000007ffffe000000000000000000000000001ffffe8000000000000000000000000007ffff8000000000000000000000000002ffffe000000000000000000000000000ffffe8000000000000000000000000003ffff8000000000000000000000000000fffff000000000000000000000000000fffffc000000000000000000000000003ffffc000000000000000000000000000fffff000000000000000000000000000fffff0000000000000000000000000003ffff0000000000000000000000000000ffff00000000000000000000000000005fff00000000000000000000000000003fff000000000000000000000000000007ff000000000000000000000000000001ff000000000000000000000000000001ff0000000000000000000000000000007f0000000000000000000000000000001f0000000000000000000000000000001e0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000"`;
